main.nf

```diff
@@ -2115,6 +2129,94 @@ process maltextract {
   """
 }
 
+/*
+ * Step 17-B: Metagenomic screening of unmapped reads: Kraken2
+*/
+
+if (params.run_metagenomic_screening && params.database.endsWith(".tar.gz") && params.metagenomic_tool == 'kraken'){
+  comp_kraken = file(params.database)
+
+  process decomp_kraken {
+    input:
+    file(ckdb) from comp_kraken
+    
+    output:
+    file(dbname) into ch_krakendb
+    
+    script:
+    dbname = params.database.tokenize("/")[-1].tokenize(".")[0]
+    """
+    tar xvzf $ckdb
+    """
+  }
+
+} else if (! params.database.endsWith(".tar.gz") && params.run_metagenomic_screening && params.metagenomic_tool == 'kraken') {
+    ch_krakendb = file(params.database)
+} else {
+    ch_krakendb = Channel.empty()
+}
+
+
+process kraken {
+  tag "$prefix"
+  label 'mc_huge'
+  publishDir "${params.outdir}/metagenomic_classification/kraken", mode:"copy"
+
+  when:
+  params.run_metagenomic_screening && params.run_bam_filtering && params.bam_discard_unmapped && params.bam_unmapped_type == 'fastq' && params.metagenomic_tool == 'kraken'
+
+  input:
+  file fastq from ch_bam_filtering_for_metagenomic_kraken
+  file(krakendb) from ch_krakendb
+
+  output:
+  file "*.kraken.out" into ch_kraken_out
+  set val(prefix), file("*.kreport") into ch_kraken_report
+  
+  
+  script:
+  prefix = fastq.toString().tokenize('.')[0]
+  out = prefix+".kraken.out"
+  kreport = prefix+".kreport"
+
+  """
+  kraken2 --db ${krakendb} --threads ${task.cpus} --output $out --report $kreport $fastq
```

Comment:
Any reason why the following flags aren't included? 
```
FASTQ input: Input is normally expected to be in FASTA format, but you can classify FASTQ data using the --fastq-input switch.

Compressed input: Kraken 2 can handle gzip and bzip2 compressed files as input by specifying the proper switch of --gzip-compressed or --bzip2-compressed.

https://ccb.jhu.edu/software/kraken2/index.shtml?t=manual#classification
```

Does kraken2 do auto detection too?

Reply:
> Input format auto-detection: If regular files are specified on the command line as input, Kraken 2 will attempt to determine the format of your input prior to classification. You can disable this by explicitly specifying --fasta-input, --fastq-input, --gzip-compressed, and/or --bzip2-compressed as appropriate. Note that use of the character device file /dev/fd/0 to read from standard input (aka stdin) will not allow auto-detection.

Reply:
lol. Good job reading James...